File: app/src/main/java/au/com/codeka/weather/WeatherMapFragment.java
package au.com.codeka.weather;

import android.app.Activity;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import com.github.florent37.materialviewpager.MaterialViewPagerHelper;
import com.github.ksoichiro.android.observablescrollview.ObservableScrollView;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.BitmapDescriptor;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.GroundOverlay;
import com.google.android.gms.maps.model.GroundOverlayOptions;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

import au.com.codeka.weather.model.WeatherInfo;

/**
 * Fragment which displays a map with some weather info overlayed over it.
 */
public class WeatherMapFragment extends Fragment {
  private FrameLayout refreshButton;
  private ObservableScrollView scrollView;
  private SupportMapFragment mapFragment;
  private GoogleMap map;

  private Handler handler = new Handler();
  private boolean firstTimeVisible = true;
  private boolean needsRefresh;

  private boolean overlayLoading;
  private GroundOverlay overlay;
  private LatLngBounds overlayLatLngBounds;
  private boolean needsReposition;
  private List<Bitmap> overlayBitmaps;
  private int overlayFrame;

  @Override
  public View onCreateView(
      LayoutInflater inflater,
      ViewGroup container,
      Bundle savedInstanceState) {
    View rootView = inflater.inflate(R.layout.weather_map_layout, container, false);
    scrollView = (ObservableScrollView) rootView.findViewById(R.id.scroll_view);
    refreshButton = (FrameLayout) rootView.findViewById(R.id.refresh_button);

    mapFragment = new SupportMapFragment();
    FragmentTransaction fragmentTransaction = getChildFragmentManager().beginTransaction();
    fragmentTransaction.replace(R.id.map, mapFragment);
    fragmentTransaction.commit();

    refreshButton.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View v) {
        refreshOverlay();
      }
    });
    refreshButton.setClickable(false);

    mapFragment.getMapAsync(new OnMapReadyCallback() {
      @Override
      public void onMapReady(GoogleMap googleMap) {
        // Start off at the weather location
        WeatherInfo weatherInfo = WeatherManager.i.getCurrentWeather(getActivity());
        if (weatherInfo == null) {
          // TODO: refresh once it loads
          return;
        }
        LatLng latlng = new LatLng(weatherInfo.getLat(), weatherInfo.getLng());
        googleMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latlng, 8.0f));

        map = googleMap;

        if (needsRefresh) {
          refreshOverlay();
        }
        needsRefresh = false;
      }
    });

    return rootView;
  }

  @Override
  public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
    MaterialViewPagerHelper.registerScrollView(getActivity(), scrollView, null);
  }

  /**
   * Called by the framework when we're visible/invisible to the user. The first time we're visible
   * is when we actually want to start loading the weather overlay.
   */
  @Override
  public void setUserVisibleHint(boolean isVisibleToUser) {
    super.setUserVisibleHint(isVisibleToUser);

    if (isVisibleToUser && firstTimeVisible) {
      refreshOverlay();
      firstTimeVisible = false;
    }
  }

  /**
   * Refreshes the overlay so that it's displaying on whatever the map is currently looking at.
   * Must run on the UI thread.
   */
  private void refreshOverlay() {
    if (map == null) {
      needsRefresh = true;
      return;
    }

    overlayLatLngBounds = map.getProjection().getVisibleRegion().latLngBounds;
    new Thread(overlayUpdateRunnable).start();
  }

  private final Runnable overlayUpdateRunnable = new Runnable() {
    @Override
    public void run() {
      overlayLoading = true;
      getActivity().runOnUiThread(updateRefreshButtonRunnable);

      InputStream ins = WeatherManager.i.fetchMapOverlay(
          overlayLatLngBounds,
          mapFragment.getView().getWidth() / 2,
          mapFragment.getView().getHeight() / 2);
      if (ins == null) {
        return;
      }

      GifDecoder gifDecoder = new GifDecoder();
      gifDecoder.read(ins);

      overlayBitmaps = new ArrayList<>();
      for (int frame = 0; frame < gifDecoder.getFrameCount(); frame++) {
        overlayBitmaps.add(gifDecoder.getFrame(frame));
      }

      overlayLoading = false;
      needsReposition = true;

      // could be null if we've finished by the time this runs.
      Activity activity = getActivity();
      if (activity != null) {
        activity.runOnUiThread(updateOverlayFrameRunnable);
      }
    }
  };

  private final Runnable updateOverlayFrameRunnable = new Runnable() {
    @Override
    public void run() {
      BitmapDescriptor bitmapDescriptor =
          BitmapDescriptorFactory.fromBitmap(overlayBitmaps.get(overlayFrame));
      if (overlay == null) {
        overlay = map.addGroundOverlay(new GroundOverlayOptions()
            .positionFromBounds(overlayLatLngBounds)
            .image(bitmapDescriptor));
      } else {
        overlay.setImage(bitmapDescriptor);
      }

      if (needsReposition) {
        overlay.setPositionFromBounds(overlayLatLngBounds);
        needsReposition = false;
      }

      overlayFrame++;
      if (overlayFrame >= overlayBitmaps.size()) {
        overlayFrame = 0;
      }

      // replace the existing callback (if there is one) with a new one.
      handler.removeCallbacks(updateOverlayFrameRunnable);
      handler.postDelayed(updateOverlayFrameRunnable, 1000);

      updateRefreshButtonRunnable.run();
    }
  };

  private final Runnable updateRefreshButtonRunnable = new Runnable() {
    @Override
    public void run() {
      if (overlayLoading) {
        refreshButton.setClickable(false);
        refreshButton.findViewById(R.id.refresh_progress).setVisibility(View.VISIBLE);
        refreshButton.findViewById(R.id.refresh_text).setVisibility(View.GONE);
      } else {
        refreshButton.setClickable(true);
        refreshButton.findViewById(R.id.refresh_progress).setVisibility(View.GONE);
        refreshButton.findViewById(R.id.refresh_text).setVisibility(View.VISIBLE);
      }
    }
  };
}

File: app/src/main/java/au/com/codeka/weather/WeatherMapFragment.kt
package au.com.codeka.weather

import android.app.Activity
import android.graphics.Bitmap
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.core.widget.NestedScrollView
import androidx.fragment.app.Fragment
import au.com.codeka.weather.model.MapOverlay
import au.com.codeka.weather.model.WeatherInfo
import com.github.florent37.materialviewpager.MaterialViewPagerHelper
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.*
import java.io.InputStream
import java.util.*

/**
 * Fragment which displays a map with some weather info overlayed over it.
 */
class WeatherMapFragment : Fragment() {
  companion object {
    private const val TAG = "codeka.weather"
  }

  private var refreshButton: FrameLayout? = null
  private var scrollView: NestedScrollView? = null
  private var mapFragment: SupportMapFragment? = null
  private var map: GoogleMap? = null
  private val handler = Handler()
  private var firstTimeVisible = true
  private var needsRefresh = false
  private var overlayLoading = false
  private var overlay: GroundOverlay? = null
  private var overlayLatLngBounds: LatLngBounds? = null
  private var needsReposition = false
  private var mapOverlays: ArrayList<MapOverlay> = arrayListOf()
  private var overlayFrame = 0

  override fun onCreateView(
      inflater: LayoutInflater,
      container: ViewGroup?,
      savedInstanceState: Bundle?): View? {
    val rootView = inflater.inflate(R.layout.weather_map_layout, container, false)
    scrollView = rootView.findViewById(R.id.scroll_view)
    refreshButton = rootView.findViewById(R.id.refresh_button)
    mapFragment = SupportMapFragment()
    val fragmentTransaction = childFragmentManager.beginTransaction()
    fragmentTransaction.replace(R.id.map, mapFragment!!)
    fragmentTransaction.commit()
    refreshButton!!.setOnClickListener { refreshOverlay() }
    refreshButton!!.isClickable = false
    mapFragment!!.getMapAsync(OnMapReadyCallback { googleMap -> // Start off at the weather location
      val weatherInfo: WeatherInfo = WeatherManager.i.getCurrentWeather(activity)
          ?: // TODO: refresh once it loads
          return@OnMapReadyCallback
      val latlng = LatLng(weatherInfo.lat, weatherInfo.lng)
      googleMap.moveCamera(CameraUpdateFactory.newLatLngZoom(latlng, 8.0f))
      map = googleMap
      if (needsRefresh) {
        refreshOverlay()
      }
      needsRefresh = false
    })
    return rootView
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    MaterialViewPagerHelper.registerScrollView(context, scrollView)
  }

  /**
   * Called by the framework when we're visible/invisible to the user. The first time we're visible
   * is when we actually want to start loading the weather overlay.
   */
  override fun setUserVisibleHint(isVisibleToUser: Boolean) {
    super.setUserVisibleHint(isVisibleToUser)
    if (isVisibleToUser && firstTimeVisible) {
      refreshOverlay()
      firstTimeVisible = false
    }
  }

  /**
   * Refreshes the overlay so that it's displaying on whatever the map is currently looking at.
   * Must run on the UI thread.
   */
  private fun refreshOverlay() {
    if (map == null) {
      needsRefresh = true
      return
    }

    overlayLatLngBounds = map!!.projection.visibleRegion.latLngBounds
    Thread(overlayUpdateRunnable).start()
  }

  private val overlayUpdateRunnable = Runnable {
    overlayLoading = true
    activity!!.runOnUiThread(updateRefreshButtonRunnable)
    mapOverlays = WeatherManager.i.fetchMapOverlay(
        overlayLatLngBounds!!,
        mapFragment!!.view!!.width / 2,
        mapFragment!!.view!!.height / 2)
        ?: return@Runnable
    overlayLoading = false
    needsReposition = true

    // could be null if we've finished by the time this runs.
    val activity: Activity? = activity
    activity?.runOnUiThread(updateOverlayFrameRunnable)
  }

  private val updateOverlayFrameRunnable: Runnable = object : Runnable {
    override fun run() {
      if (mapOverlays.size == 0) {
        return
      }
      val bitmapDescriptor = BitmapDescriptorFactory.fromBitmap(mapOverlays[overlayFrame].bitmap)
      if (overlay == null) {
        val overlayOptions = GroundOverlayOptions()
            .image(bitmapDescriptor)
        if (mapOverlays[overlayFrame].latLngBounds != null) {
          overlayOptions.positionFromBounds(overlayLatLngBounds)
        } else {
          Log.i(TAG, "fitting to latlng + widthMeters")
          overlayOptions.position(
              mapOverlays[overlayFrame].latLng, mapOverlays[overlayFrame].widthMeters!!)
        }
        overlay = map!!.addGroundOverlay(overlayOptions)
      } else {
        overlay!!.setImage(bitmapDescriptor)
      }
      if (needsReposition) {
        overlay!!.setPositionFromBounds(overlayLatLngBounds)
        needsReposition = false
      }
      overlayFrame++
      if (overlayFrame >= mapOverlays.size) {
        overlayFrame = 0
      }

      // replace the existing callback (if there is one) with a new one.
      handler.removeCallbacks(this)
      handler.postDelayed(this, 1000)
      updateRefreshButtonRunnable.run()
    }
  }

  private val updateRefreshButtonRunnable = Runnable {
    if (overlayLoading) {
      refreshButton!!.isClickable = false
      refreshButton!!.findViewById<View>(R.id.refresh_progress).visibility = View.VISIBLE
      refreshButton!!.findViewById<View>(R.id.refresh_text).visibility = View.GONE
    } else {
      refreshButton!!.isClickable = true
      refreshButton!!.findViewById<View>(R.id.refresh_progress).visibility = View.GONE
      refreshButton!!.findViewById<View>(R.id.refresh_text).visibility = View.VISIBLE
    }
  }
}